lmcache/v1/storage_backend/connector/mooncakestore_connector.py

```
@@ -213,19 +426,69 @@ async def get(self, key: CacheEngineKey) -> Optional[MemoryObj]:
             return None
 
     async def put(self, key: CacheEngineKey, memory_obj: MemoryObj):
-        # Please use a function like `memory_obj.to_meta()`.
-        kv_bytes = memory_obj.byte_array
-        kv_shape = memory_obj.get_shape()
-        kv_dtype = memory_obj.get_dtype()
-        memory_format = memory_obj.get_memory_format()
-
-        metadata_bytes = RemoteMetadata(
-            len(kv_bytes), kv_shape, kv_dtype, memory_format
-        ).serialize()
-        assert len(metadata_bytes) == METADATA_BYTES_LEN
+        """
+        Put operation with metadata-consistent handling.
+        Uses put_from (without metadata) or
+        put_parts (with metadata) to match get behavior.
+        """
         key_str = key.to_string()
 
+        # Check metadata handling mode to match get behavior
+        if self.save_chunk_meta:
+            # Use put_parts with metadata stored remotely
+            await self._put_with_metadata(key_str, memory_obj)
+        else:
+            # Use put_from without metadata (zero-copy)
+            await self._put_without_metadata(key_str, memory_obj)
+
+    async def _put_without_metadata(self, key_str: str, memory_obj: MemoryObj):
+        """
+        Zero-copy put using put_from when metadata is not stored remotely.
+        This is used when save_chunk_meta=False (matches _batch_get_into).
+        """
+        try:
+            buffer_ptr = memory_obj.tensor.data_ptr()
+            buffer_size = memory_obj.tensor.numel() * memory_obj.tensor.element_size()
+
+            await asyncio.wait_for(
+                asyncio.to_thread(
+                    self.store.put_from,
+                    key_str,
+                    buffer_ptr,
+                    buffer_size,
+                    self.replica_config,
+                ),
+                timeout=self.config.transfer_timeout,
+            )
+        except asyncio.TimeoutError:
+            logger.warning(
```

I guess we need to raise the exception so that caller can aware and handling.